### ISSUE TITLE
[clang][analyzer] Pointer cast after in-bounds arithmetic loses non-null constraint

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#220281]
**Description:**
Analyzer loses the non-null constraint when a pointer produced by in-bounds pointer arithmetic is cast to another object-pointer type. It correctly proves that the source pointer is non-null, but treats the cast result as both null and non-null. Because of that it explores an infeasible branch and emits a false `core.NullDereference` warning.

**Reproducer:**
```c
void clang_analyzer_eval(int);

void f(_Bool i)
{
    unsigned char a[2];
    unsigned char *q = a + i;
    char *r = (char *)q;

    clang_analyzer_eval(q != 0); // TRUE
    clang_analyzer_eval(r != 0); // expected TRUE; actual TRUE and FALSE

    if (!r)
        *r = 0; // false positive
}
```

Both expressions should evaluate only to `TRUE`, and no null-dereference warning should be emitted. Because `i` has type `_Bool`, it is either zero or one. Therefore `q` points to either `a[0]` or `a[1]` and is non-null. Converting this properly aligned non-null object pointer from `unsigned char *` to `char *` shouldn't produce a null pointer.

Slightly complex example for which I think is related can be found [here](https://compiler-explorer.com/z/K77oK6753).

**Comment by Snape3058** [comment]
`unsigned char *q = a + i;` works normally with `q` assigned an element region: `&Element{a,reg_$1<_Bool i>,unsigned char}`.
`char *r = (char *)q;` creates a conjured symbol for `r`: `&SymRegion{conj_$3{char *, LC1, S894, #1}}`.

In `ExprEngine::handleLValueBitCast`, `SValBuilder::evalCast` returns `unknown`, and the following `escapeValues` makes `r` get this conjured symbol.

**Comment by bozicrHT** [comment]
Thanks! The cast result shouldn’t become an unrelated conjured symbol. Could we preserve the original region by wrapping it in a zero-offset ElementRegion of the target type, something like: `&Element{Element{a,reg_$1<_Bool i>,unsigned char},0 S64b,char}`?
